### DISCUSSION TITLE
Shared JDK for Users

### DISCUSSION THREAD

**Discussion by DMATVON** [microsoft/openjdk#604] [Q&A] [answered]
Hi Everyone
So far, our users have their own JDK/JRE installed on their local machine to enable them executing a Java program. There is a discussion about mounting a network drive with a shared folder in which the OpenJDK17 from Microsoft is placed, instead of having the different local installations. The idea is to change the JAVA_HOME variable pointing to the shared folder.

Does it no longer need to be installed? If yes, where does the limitations or downsides become evident of having a shared non-installed Java runtime environment?
Many thanks for sharing your thoughts on it?

**Comment by karianna** [comment, marked as answer]
Having multiple users share a single installation of the Java Development Kit (JDK) or Java Runtime Environment (JRE) can lead to several disadvantages and potential issues:

1. **Configuration Conflicts**: Different projects or users may require different JDK or JRE settings. Shared configurations can lead to conflicts, such as differences in environmental variables like `JAVA_HOME` or `PATH`. Adjustments by one user can inadvertently affect others, leading to build failures or runtime errors.

2. **Versioning Issues**: If a project needs a specific version of the JDK or JRE that is different from what another project uses, it can lead to compatibility issues. While tools like SDKMAN or containerization can help manage multiple versions on the same machine, a shared setup might not always accommodate individual project needs.

3. **Security Risks**: Shared environments increase the risk of security breaches. If one user's account is compromised, the attacker can potentially affect all projects and users that rely on the shared JDK or JRE installation. Additionally, updates or patches might be delayed if they are incompatible with the needs of certain projects, leaving vulnerabilities unaddressed.

4. **Performance Impact**: Multiple users working on the same system can strain its resources, leading to slower build times and decreased performance. This issue can be exacerbated if the projects are resource-intensive.

5. **Dependency Management**: Different projects may require different third-party libraries or tools. While this is more directly related to project dependencies than the JDK/JRE itself, a shared environment can complicate the management of these dependencies, especially if they are globally installed.

6. **Access Control**: Properly managing access to a shared JDK/JRE installation can be challenging. Ensuring that users have the necessary permissions to start their applications without inadvertently giving them the ability to affect others' work requires careful configuration of user privileges.

7. **Customization Limitations**: Users might need to customize their development environment with specific tools or plugins. In a shared environment, such customizations could interfere with other users' setups or be outright impossible due to restrictions on software installation or configuration changes.

Addressing these disadvantages often involves adopting best practices for development environment management, such as using containerization technologies (e.g., Docker) to isolate projects and their dependencies, or employing version management tools to allow users to easily switch between different JDK or JRE versions as required by their projects.

**Reply by karianna** [reply]
I'll also add that the recommended practice for Java 17+ is to try to bundle the JRE components you need into the application that you are deploying using a combination of `jlink` and `jpackage`.  Not sure if you're able to explore that space.

**Reply by DMATVON** [reply]
Hi,
I appreciate your detailed advice about that topic. 
Many thanks
